File: frontend/types/changeRequest.ts
import { ImpactReport, RiskLevel, MigrationResult } from './migration';

export type { RiskLevel, ImpactReport, MigrationResult };

// Backend: Namines.Core/Enums/ChangeRequestStatus.cs
export type ChangeRequestStatus = 'PendingReview' | 'Approved' | 'Rejected';

// Backend: Namines.Core/Enums/ApprovalDecision.cs
export type ApprovalDecision = 'Approved' | 'Rejected';

export interface ChangeRequestSummary {
  id: string;
  title: string | null;
  status: ChangeRequestStatus;
  riskLevel: RiskLevel;
  branchName: string;
  tableCount: number;
  createdAt: string;
  resolvedAt: string | null;
  approvedCount: number;
  requiredApprovals: number;
}

export interface ChangeRequestApprovalDto {
  id: string;
  userId: string;
  username: string | null;
  decision: ApprovalDecision;
  comment: string | null;
  createdAt: string;
}

export interface SchemaVersionSummary {
  id: string;
  version: number;
  tableCount: number;
  createdAt: string;
}

// Backend: Namines.Core/Models/TestRunResult.cs — "Run Tests" (G12)
export interface TestRunInfo {
  testRunSupported: boolean | null;
  testRunSuccess: boolean | null;
  testRunMessage: string | null;
  testRunFailedStatement: string | null;
  testRunDurationMs: number | null;
  testRunAt: string | null;
}

// Backend: Namines.Core/Models/AffectedCodeMatch.cs — "Affected Code" scan (G13)
export interface AffectedCodeMatch {
  fileName: string;
  lineNumber: number;
  matchedIdentifier: string;
  lineText: string;
}

export interface AffectedCodeScanResult {
  candidateIdentifiers: string[];
  matches: AffectedCodeMatch[];
  filesScanned: number;
}

// Backend: Namines.Core/Models/Auth/ChangeRequestAuditLog.cs — G16
export type ChangeRequestAuditAction = 'Created' | 'AutoApproved' | 'Approved' | 'Rejected';

export interface ChangeRequestAuditEntry {
  id: string;
  action: ChangeRequestAuditAction;
  actorUserId: string | null;
  actorUsername: string | null;
  details: string | null;
  createdAt: string;
}

export interface ChangeRequestDetail {
  id: string;
  projectId: string;
  branchId: string;
  branchName: string;
  title: string | null;
  status: ChangeRequestStatus;
  riskLevel: RiskLevel;
  createdByUserId: string;
  createdAt: string;
  resolvedAt: string | null;
  headVersion: SchemaVersionSummary;
  baseVersion: SchemaVersionSummary | null;
  impact: ImpactReport;
  aiExplanation: string | null;
  migration: MigrationResult | null;
  testRun: TestRunInfo | null;
  requiredApprovals: number;
  approvedCount: number;
  rejectedCount: number;
  approvals: ChangeRequestApprovalDto[];
}

File: backend/Namines.Core/Models/Auth/ChangeRequest.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using Namines.Core.Enums;

namespace Namines.Core.Models.Auth
{
    /// <summary>
    /// "Database PR" — new-phase/29-DATABASE-CHANGE-REVIEW.md. İki <see cref="SchemaVersion"/>
    /// arasındaki <see cref="Analysis.SchemaImpactAnalyzer"/> raporunu, insan onayı gerektiren
    /// kalıcı bir kayda dönüştürür. G10'un (Branch/SchemaVersion) üzerine oturur.
    ///
    /// Kapsam notu: doc'taki tam yaşam döngüsünün (draft/analyzing/applying/applied/failed)
    /// yalnızca review kısmı burada — apply pipeline'ı yok (bkz. ChangeRequestStatus).
    /// </summary>
    public class ChangeRequest
    {
        [Key]
        public string Id { get; set; } = Guid.NewGuid().ToString();

        public string ProjectId { get; set; } = null!;
        public CloudProject Project { get; set; } = null!;

        public string BranchId { get; set; } = null!;
        public Branch Branch { get; set; } = null!;

        /// <summary>Null ise "boş şemadan" karşılaştırılıyor demektir (branch'in ilk versiyonu).</summary>
        public string? BaseVersionId { get; set; }
        public SchemaVersion? BaseVersion { get; set; }

        public string HeadVersionId { get; set; } = null!;
        public SchemaVersion HeadVersion { get; set; } = null!;

        public string? Title { get; set; }

        public ChangeRequestStatus Status { get; set; } = ChangeRequestStatus.PendingReview;

        /// <summary><see cref="Models.ImpactReport.OverallRisk"/> ile aynı değer — onay
        /// gereksinimini hesaplamak için ayrı bir sorgu/deserileştirme gerektirmesin diye kopyalandı.</summary>
        public RiskLevel RiskLevel { get; set; }

        /// <summary>Oluşturma anında hesaplanmış <see cref="Models.ImpactReport"/>'un JSON hâli —
        /// deterministik olduğu için tekrar hesaplamaya gerek yok, ama drift'e karşı saklanır.</summary>
        public string ImpactReportJson { get; set; } = null!;

        public string CreatedByUserId { get; set; } = null!;
        public ApplicationUser CreatedByUser { get; set; } = null!;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? ResolvedAt { get; set; }

        // ── G12 — "Run Tests" sonucu (new-phase/29 §4) ──────────────────────
        // Impact Analysis bir TAHMİN; bu alanlar üretilen DDL'in gerçek, ephemeral bir
        // motor container'ında ÇALIŞTIRILDIĞININ kanıtıdır. Null = hiç çalıştırılmadı.
        public bool? TestRunSuccess { get; set; }
        public bool? TestRunSupported { get; set; }
        public string? TestRunMessage { get; set; }
        public string? TestRunFailedStatement { get; set; }
        public long? TestRunDurationMs { get; set; }
        public DateTime? TestRunAt { get; set; }

        public List<ChangeRequestApproval> Approvals { get; set; } = new();
    }

    /// <summary>
    /// Bir kullanıcının bir <see cref="ChangeRequest"/> üzerindeki tek kararı.
    /// new-phase/29-DATABASE-CHANGE-REVIEW.md §3: risk seviyesine göre 1 veya 2 farklı
    /// kişinin onayı gerekir; Destructive/Breaking'de onaylayan, değişikliği yapan
    /// kişiyle aynı olamaz (uygulama katmanında <c>ChangeRequestController</c>'da denetlenir).
    /// </summary>
    public class ChangeRequestApproval
    {
        [Key]
        public string Id { get; set; } = Guid.NewGuid().ToString();

        public string ChangeRequestId { get; set; } = null!;
        public ChangeRequest ChangeRequest { get; set; } = null!;

        public string UserId { get; set; } = null!;
        public ApplicationUser User { get; set; } = null!;

        public ApprovalDecision Decision { get; set; }
        public string? Comment { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }
}
